sdk/tests/invariant.test.ts

```diff
@@ -0,0 +1,255 @@
+      feeGrowthOutsideX: { v: 0n },
+      feeGrowthOutsideY: { v: 0n },
+      secondsOutside: lowerTick.secondsOutside
+    })
+    chai.assert.deepEqual(upperTick, {
+      index: 10n,
+      sign: false,
+      liquidityChange: liquidityDelta,
+      liquidityGross: liquidityDelta,
+      sqrtPrice: { v: 1000500100010000000000000n },
+      feeGrowthOutsideX: { v: 0n },
+      feeGrowthOutsideY: { v: 0n },
+      secondsOutside: upperTick.secondsOutside
+    })
+  })
+  it('create pool', async () => {
+    const chai = await loadChai()
+    const wasm = await loadWasm()
+
+    const invariant = await Invariant.load(client, hashes.invariant.loadHash, network)
+    const feeTier = await callWasm(wasm.newFeeTier, { v: 6000000000n }, 10n)
+    const poolKey = await callWasm(
+      wasm.newPoolKey,
+      hashes.tokenX.packageHash,
+      hashes.tokenY.packageHash,
+      feeTier
+    )
+
+    await invariant.addFeeTier(deployer, feeTier)
+    const feeTierExist = await invariant.feeTierExist(feeTier)
+    chai.assert.exists(feeTierExist)
+
+    const initSqrtPrice = { v: 1000000000000000000000000n }
+    await invariant.createPool(deployer, poolKey, initSqrtPrice)
+
+    const pools = await invariant.getPools()
+    const pool = await invariant.getPool(poolKey)
+    const expectedPool = {
+      liquidity: { v: 0n },
+      sqrtPrice: { v: 1000000000000000000000000n },
+      currentTickIndex: 0n,
+      feeGrowthGlobalX: { v: 0n },
+      feeGrowthGlobalY: { v: 0n },
+      feeProtocolTokenX: { v: 0n },
+      feeProtocolTokenY: { v: 0n },
+      startTimestamp: pool.startTimestamp,
+      lastTimestamp: pool.lastTimestamp,
+      feeReceiver: pool.feeReceiver,
+      oracleInitialized: false
+    }
+
+    chai.assert.exists(pools.length === 1)
+    chai.assert.deepEqual(pool, expectedPool)
+  })
+  it('attempt to create pool with wrong tick & sqrtPrice relationship', async () => {
+    const chai = await loadChai()
+    const wasm = await loadWasm()
+
+    const invariant = await Invariant.load(client, hashes.invariant.loadHash, network)
+    const feeTier = await callWasm(wasm.newFeeTier, { v: 6000000000n }, 10n)
+    const poolKey = await callWasm(
+      wasm.newPoolKey,
+      hashes.tokenX.packageHash,
+      hashes.tokenY.packageHash,
+      feeTier
+    )
+
+    await invariant.addFeeTier(deployer, feeTier)
+    const feeTierExist = await invariant.feeTierExist(feeTier)
+    chai.assert.exists(feeTierExist)
+
+    const initSqrtPrice = { v: 1000175003749000000000000n }
+    assertThrowsAsync(
+      invariant.createPool(deployer, poolKey, initSqrtPrice),
+      wasm.InvariantError.InvalidInitTick
+    )
+  })
+  it('create pool x/y and y/x', async () => {
+    const chai = await loadChai()
+    const wasm = await loadWasm()
+
+    const invariant = await Invariant.load(client, hashes.invariant.loadHash, network)
+    const feeTier = await callWasm(wasm.newFeeTier, { v: 6000000000n }, 10n)
+    const poolKey = await callWasm(
```

Comment:
There are still many `callWasm` invocations in test to create `FeeTier` and `PoolKey`